### PR TITLE
Add Matrix4 and Quaternion to Core; rebuild Transform on native math

### PR DESCRIPTION
**Pull request by jayrulez**
Introduce core.math.matrix4 (4x4 row-major, row-vector convention: identity/translation/scale/rotation, perspective/ortho/lookAt RH, multiply, transpose,determinant, transformPoint/Direction/Point2D, and a scale-aware tryInverse/inverse) and core.math.quaternion (fromAxisAngle, Hamilton product, conjugate/dot/inverse/normalize/rotateVector/slerp, rotationMatrix), with the Matrix4 inverse bodies kept in Matrix4.cpp so the interface stays light.

Rewrite Transform to position / rotation (Quaternion) / scale with an inline toMatrix() returning Matrix4 and a lerp helper, removing the last bx dependency from Core. Update its two callers (Renderer, RenderSample) to the new API.

Add the math helpers these types need (sin/cos/tan/acos/sqrt, scalar and vector nearlyEqual/nearlyZero) and unit tests for Matrix4/Quaternion/Transform.

Review follow-ups: expose singular inverses through tryInverse with a Hadamard row-norm threshold instead of a silent identity; guard a zero-length axis in fromAxisAngle; treat antipodal quaternions as equal; make the near-equal helpers constexpr and drop the redundant approx_eq; use the multidimensional subscript operator on Matrix4.

TypesCommon zero/one: left as-is - they're static members; the out-of-line Vector3::zero definition requires the in-class declaration

PIC: Runtime is the only shared library, and it links the whole Core closure (Definitions/Math/IO/Memory/Logging) transitively - so all of those objects end up in libRuntime.so and must be PIC, not just Math. the only consumers paying the PIC cost otherwise are the test exes, which aren't perf-sensitive.

## Summary by CodeRabbit

* **New Features**
  * Added new C++ math modules for 4x4 matrices and quaternions, including projection/view helpers and quaternion rotation utilities.
  * Added scale-aware matrix inversion (`tryInverse`) plus a debug-checked `inverse`, and `Transform::toMatrix` / `Transform::lerp`.
  * Introduced `nearlyEqual` helpers across vectors and math functions, plus `Vector3::zero()` / `Vector3::one()`.

* **Bug Fixes**
  * Improved robustness for singular/near-zero inversion and quaternion construction/normalization edge cases.

* **Tests**
  * Expanded matrix, quaternion, transform, and numerical edge-case validation.

### CODE REVIEW

File: Engine/cpp/Runtime/Rendering/Renderer/Renderer.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@ module rendering.renderer;
 
 import core.stdtypes;
 import core.math.transform;
+import core.math.matrix4;
```
Comment (OldDev78) [inline Engine/cpp/Runtime/Rendering/Renderer/Renderer.cpp]:
```suggestion
import core;
```

File: Engine/cpp/Runtime/Rendering/Renderer/Renderer.cpp [hunk 2]
```diff
@@ -108,7 +109,8 @@ namespace draco::rendering::renderer
 
         buildUniforms(material, p.uniforms);
 
-        transform.toMatrix(p.model);
+        const draco::math::Matrix4 model = transform.toMatrix();
+        std::memcpy(p.model, model.data(), sizeof(p.model));
```
Comment (OldDev78) [inline Engine/cpp/Runtime/Rendering/Renderer/Renderer.cpp]: Requiring a construction of a temporary is a step backwards compared to the previous API. Either allow assigning a `Matrix4` to `p.model`, or restore the array filling function as an overload.

File: Samples/cpp/Render/RenderSample.cpp
Comment (OldDev78) [inline Samples/cpp/Render/RenderSample.cpp:1]: It's ok to do `using namespace draco;` and/or `using namespace draco::math;` inside the body of `main`.

File: Engine/cpp/Runtime/Core/Math/TypesCommon.cppm [hunk 3]
```diff
@@ -63,6 +63,10 @@ export namespace draco::math {
         [[nodiscard]] static Vector3 spherical(f32 azimuth, f32 inclination, f32 radius = 1.0f) noexcept;
         [[nodiscard]] static Vector3 cylindrical(f32 angle, f32 radius = 1.0f, f32 height = 0.0f) noexcept;
 
+        // constant vectors (defined in Vector3.cppm)
+        static const Vector3 zero;
```
Comment (jayrulez) [inline Engine/cpp/Runtime/Core/Math/TypesCommon.cppm]: Can't make these inline constexpr at the declaration. they're static members of Vector3's own incomplete type, and its constexpr ctor lives in Vector3.cppm which imports this partition, so it isn't reachable here. The out-of-line inline constexpr definition already makes them usable in constexpr contexts; Transform now uses direct construction regardless.

Comment (OldDev78) [reply Engine/cpp/Runtime/Core/Math/TypesCommon.cppm]: This is a non-conformant behavior of clang. This code does not compile on g++. The `one` and `zero` you declared are runtime statics - not compile time. It's true that you cannot simply declare them `static constexpr` here, but promoting qualifier from `static const` in the declaration to `inline constexpr` in the definition is incorrect behavior, and clang should not have accepted this.

Declare them as functions instead.

```
static constexpr Vector3 zero();
static constexpr Vector3 one();
```

File: Engine/cpp/Runtime/Core/Math/Transform.cppm [hunk 4]
```diff
@@ -1,70 +1,44 @@
...
+import core.math.types;
+import core.math.matrix4;
+import core.math.quaternion;
 
 export namespace draco::math
 {
+    // =======================================================================
+    // Transform - position / rotation / scale, composed as S * R * T.
+    // =======================================================================
     struct Transform
```
Comment (OldDev78) [inline Engine/cpp/Runtime/Core/Math/Transform.cppm]: It's easier to declare the struct itself `struct [[nodiscard]] Transform`. Then any function returning this type is nodiscard. Same for any of the math types.

File: Engine/cpp/Runtime/Core/Math/Transform.cppm [hunk 5]
```diff
@@ -1,70 +1,44 @@
...
-            return Vector3{scale[0], scale[1], scale[2]};
-        }
-
-        // Compute column-major matrix from transform
-        void toMatrix(f32 out[16]) const;
     };
+
+    // Identity transform (position 0, rotation identity, scale 1) - the
+    // default-constructed value.
+    inline constexpr Transform identityTransform{};
```
Comment (OldDev78) [inline Engine/cpp/Runtime/Core/Math/Transform.cppm]: Please declare `static constexpr`, or as a function `identity()` (no repetition of Transform, right now one has to type `Transform::identityTransform`).

This case worked, because this type is aggregate and complete at this point, unlike what happened with the Vector types.

**Review by OldDev78 — CHANGES_REQUESTED** [review]

File: Engine/cpp/Runtime/Core/CMakeLists.txt [hunk 6]
```diff
@@ -1,12 +1,13 @@
-add_modules_library(Definitions)
-add_modules_library(Math)
-add_modules_library(IO)
-add_modules_library(Memory)
-add_modules_library(Logging)
-add_modules_library(Containers)
+# PIC: these are linked into the shared Runtime library (via Core), so their objects
+# must be position-independent - a non-PIC static (e.g. Quaternion::identity) otherwise
+# fails to relocate when making the shared object.
```
Comment (OldDev78) [inline Engine/cpp/Runtime/Core/CMakeLists.txt]: This propagates **up** from Math to Core, not sideways to the others. The Definitions library already needed PIC before iirc. Only use PIC when strictly necessary.

**Review by OldDev78 — DISMISSED** [review]
Make sure you double check how PIC is declared, don´t just flag everything inside Core, just because Core needs it. Only declare individual static libraries as position independent code if strictly necessary.

Otherwise, fine.
